### PR TITLE
AIP-84 Get Configs / Get Config Value

### LINKED ISSUE #42745: AIP-84 Migrate Config related public endpoint to FastAPI
### Description

Migrate the following endpoint:

- [x] Get Current configuration
- [x] Get a option from configuration

### Use case/motivation

[AIP-84](https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-84+UI+REST+API)

### Related issues

Relates to: https://github.com/apache/airflow/issues/42370

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by jason810496**
closes: #42745
related: #42370

### CODE REVIEW

File: airflow/api_fastapi/core_api/routes/public/config.py
Comment (jscheffl) [inline airflow/api_fastapi/core_api/routes/public/config.py]: These lines are redundant to lines 46-62 - can you push this to a common function and use this from both methods?

Comment (jason810496) [reply airflow/api_fastapi/core_api/routes/public/config.py]: Sure, just resolved. Thanks for the reminder!

File: airflow/api_fastapi/core_api/routes/public/config.py
Comment (pierrejeambrun) [inline airflow/api_fastapi/core_api/routes/public/config.py]: Nice.

Mimetype on fastapi are already defined in the `dag_source.py` file. What we need to do is extract from `dag_source.py` and this file the `mimetype literal` or enum, and put that in a common file, most likely `common/types.py`.

As you will see we also need the wildcard mimetype `*/*` because some http client send that to mention that they accept everything. In that case we should most likely default to `application/json`.

And then re-use that literal in both here and `dag_sources`.

Lastly you can use the `Annotated` notation, which is more modern and `Annotated[MimeTypeLiteral, Header()]`. (Not sure if the ellipsis `...` is needed).

And remove the `NOT_ACCEPTABLE` code path for the `dag_source`. (handled by pydantic validation now)

File: airflow/api_fastapi/core_api/routes/public/config.py
Comment (pierrejeambrun) [inline airflow/api_fastapi/core_api/routes/public/config.py]: You should add example responses because there is no way for FastAPI to populate correctly the openapi spec and swagger based on that.

You can take a look at how we do it in `get_dag_source`. (We combine `responses` for text/plain and `response_model` for application/json documentation)

File: airflow/api_fastapi/core_api/routes/public/config.py
Comment (pierrejeambrun) [inline airflow/api_fastapi/core_api/routes/public/config.py]: Same for docs.

File: airflow/api_fastapi/core_api/routes/public/config.py
Comment (pierrejeambrun) [reply airflow/api_fastapi/core_api/routes/public/config.py]: Also as you will see in `dag_source`. We use `start_with` because sometimes the mimetype can be `application/json; charset=utf-8` and not an exact match with `application_json`, but i didn't look to much to it.

I think some tests were failing otherwise.

File: tests/api_fastapi/core_api/routes/public/test_config.py
Comment (pierrejeambrun) [inline tests/api_fastapi/core_api/routes/public/test_config.py:270]: I think we are missing a test with `non-sensitive-only` ?

**Review by pierrejeambrun — COMMENTED** [review]
Nice.

A few suggestions but looking good.

File: tests/api_fastapi/core_api/routes/public/test_config.py
Comment (jason810496) [reply tests/api_fastapi/core_api/routes/public/test_config.py:270]: It seems the legacy API missed the `non-sensitive-only` case in the `get_config` endpoint. I’ll add this case along with a test. Thanks !

File: airflow/api_fastapi/core_api/routes/public/config.py
Comment (jason810496) [reply airflow/api_fastapi/core_api/routes/public/config.py]: Thanks for pointing that out ! 
I’ll refactor the header query into a common module.

File: tests/api_fastapi/core_api/routes/public/test_config.py
Comment (pierrejeambrun) [reply tests/api_fastapi/core_api/routes/public/test_config.py:270]: I think you can take inspiration from `test_should_respond_200_text_plain_with_non_sensitive_only` from legacy tests.

**Comment by jason810496** [comment]
Hi @pierrejeambrun, here is the update:

### Common `Accept` Header for JSON and Text
I have added `HeaderAcceptJsonOrText` in `common/headers.py`, annotated with `Mimetype` from `common/types.py`. 

I tested it with `application/json; charset=utf-8`, and for headers including `utf-8`, we need to use `startswith` to match the header correctly. Additionally, I included an OpenAPI schema for the `Accept` header.

File: airflow/api_fastapi/common/headers.py
Comment (pierrejeambrun) [inline airflow/api_fastapi/common/headers.py]: Are we missing the `Mimetype.Any` ? (maybe dynamically retrieve the `Mimetype` Enum values), or maybe it supports an enum out of the box ?

File: airflow/api_fastapi/core_api/openapi/v1-generated.yaml
Comment (pierrejeambrun) [inline airflow/api_fastapi/core_api/openapi/v1-generated.yaml]: Here we need at least 2 sections (as in the legacy doc) to show that multiples sections are returned in comparison to the other endpoint.

File: airflow/api_fastapi/core_api/openapi/v1-generated.yaml
Comment (pierrejeambrun) [inline airflow/api_fastapi/core_api/openapi/v1-generated.yaml:1913]: Why do we have line break between every lines in the example ?

Comment (pierrejeambrun) [reply airflow/api_fastapi/core_api/openapi/v1-generated.yaml:1913]: Also should we only have one value here ?

File: tests/api_fastapi/core_api/routes/public/test_config.py
Comment (pierrejeambrun) [inline tests/api_fastapi/core_api/routes/public/test_config.py:32]: We are missing a test with `*/*`  I think

File: tests/api_fastapi/core_api/routes/public/test_dag_sources.py
Comment (pierrejeambrun) [inline tests/api_fastapi/core_api/routes/public/test_dag_sources.py:78]: Here too `*/*` should behave as `application/json`

**Review by pierrejeambrun — COMMENTED** [review]
Thanks looking good, a few suggestions / question, then we can merge.

File: airflow/api_fastapi/core_api/openapi/v1-generated.yaml
Comment (jason810496) [reply airflow/api_fastapi/core_api/openapi/v1-generated.yaml:1913]: Hi @pierrejeambrun 
The OpenAPI schema at `http://localhost:29091/docs#/Config/get_config` appears to be formatted correctly:  
get_config OpenAPI schema


The line breaks are introduced by the auto-generation script (`scripts/ci/pre_commit/update_fastapi_api_spec.py`, which runs `scripts/in_container/run_update_fastapi_api_spec.py` using `get_openapi` provided by `FastAPI`).  

I prefer not to manually remove the line breaks because once this PR is merged, the line breaks will be reintroduced when others run the pre-commit script. wdyt ?

Comment (pierrejeambrun) [reply airflow/api_fastapi/core_api/openapi/v1-generated.yaml:1913]: Got it, no worry then, as long as the doc appears properly

**Review by pierrejeambrun — APPROVED** [review]
Good to merge after rebase

**Comment by jason810496** [comment]
Just resolve the conflict  🚀

**Comment by pierrejeambrun** [comment]
Hello @jason810496 do you mind fixing the formatting errors, it would be great to merge this one :)

File: tests/api_fastapi/core_api/routes/public/test_dag_sources.py
Comment (rawwar) [inline tests/api_fastapi/core_api/routes/public/test_dag_sources.py]:
```suggestion
    def test_should_respond_200_json(self, test_client, test_dag, headers):
```

Comment (rawwar) [reply tests/api_fastapi/core_api/routes/public/test_dag_sources.py:80]: @pierrejeambrun, @jason810496  , can you commit this

File: airflow/api_fastapi/core_api/routes/public/dag_sources.py
Comment (rawwar) [inline airflow/api_fastapi/core_api/routes/public/dag_sources.py]:
```suggestion
from fastapi import Depends, HTTPException, Response, status
```

Comment (rawwar) [reply airflow/api_fastapi/core_api/routes/public/dag_sources.py]: I ran pre-commits locally and ruff gave this fix

Comment (jason810496) [reply airflow/api_fastapi/core_api/routes/public/dag_sources.py]: I think it cause by my manual resolve conflict when rebasing to latest main. Just resolve, thanks !

**Comment by bbovenzi** [comment]
Screenshot 2024-11-20 at 11 32 21 AM


I was just testing this and noticed that we're passing all the values as strings. It would be nice if I am passing 'application/json' that values are actually converted so `'True'` becomes `true`, `'5'` becomes `5`

**Comment by rawwar** [comment]
> I was just testing this and noticed that we're passing all the values as strings. It would be nice if I am passing 'application/json' that values are actually converted so `'True'` becomes `true`, `'5'` becomes `5`

I think, that would require us to maintain a Model with all possible configs along with their type. From what I remember, we typecast them after retrieving from the config. For example, retries are fetched as a string and then typecased in methods.

**Comment by pierrejeambrun** [comment]
I would be for creating an issue, and tagging that as a feature request ?

**Comment by bbovenzi** [comment]
I think I'll make this a feature for the UI only config endpoint

**Pull request merged**
